network/netcore/src/transport/tcp.rs

```diff
@@ -193,36 +198,32 @@ impl AsyncWrite for TcpSocket {
     }
 }
 
-fn multiaddr_to_socketaddr(addr: &NetworkAddress) -> ::std::io::Result<SocketAddr> {
-    let mut iter = addr.as_slice().iter();
-    let proto1 = iter.next().ok_or_else(|| {
-        io::Error::new(
-            io::ErrorKind::InvalidInput,
-            format!("Invalid NetworkAddress '{:?}'", addr),
-        )
-    })?;
-    let proto2 = iter.next().ok_or_else(|| {
-        io::Error::new(
-            io::ErrorKind::InvalidInput,
-            format!("Invalid NetworkAddress '{:?}'", addr),
-        )
-    })?;
-
-    if iter.next().is_some() {
+/// parse the `NetworkAddress` into the `/ip4/<ip>/tcp/<port>` or
+/// `/ip6/<ip>/tcp/<port>` prefix and unparsed `&[Protocol]` suffix.
+fn parse_socketaddr<'a>(
+    addr: &'a NetworkAddress,
+) -> ::std::io::Result<(SocketAddr, &'a [Protocol])> {
+    if addr.len() < 2 {
         return Err(io::Error::new(
             io::ErrorKind::InvalidInput,
             format!("Invalid NetworkAddress '{:?}'", addr),
         ));
     }
 
-    match (proto1, proto2) {
-        (Protocol::Ip4(ip), Protocol::Tcp(port)) => Ok(SocketAddr::new((*ip).into(), *port)),
-        (Protocol::Ip6(ip), Protocol::Tcp(port)) => Ok(SocketAddr::new((*ip).into(), *port)),
-        _ => Err(io::Error::new(
-            io::ErrorKind::InvalidInput,
-            format!("Invalid NetworkAddress '{:?}'", addr),
-        )),
-    }
+    let (prefix, suffix) = addr.as_slice().split_at(2);
+
+    let socket_addr = match prefix {
+        [Protocol::Ip4(ip), Protocol::Tcp(port)] => SocketAddr::new((*ip).into(), *port),
+        [Protocol::Ip6(ip), Protocol::Tcp(port)] => SocketAddr::new((*ip).into(), *port),
+        _ => {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                format!("Invalid NetworkAddress '{:?}'", addr),
+            ))
+        }
+    };
+
+    Ok((socket_addr, suffix))
```

Comment:
I think in programming language this pattern is usually called a tail and not a suffix : o